### PR TITLE
Support bias in Attention

### PR DESCRIPTION
**Pull request by kakao-steve-ai**
## 🙏 Describe the pull request

* Attention with bias has been implemented.
* For now, `attn_mask` can be used for bias only.
* Issue #128 

## ✅ Checklist

- [x] Code follows the project's coding conventions and style.
- [x] Tests have been added or updated to cover the changes.
- [x] Documentation has been updated, if necessary.

### CODE REVIEW

File: tests/test_attention.py [hunk 1]
```diff
@@ -20,55 +20,73 @@
...
-        trident.function.scaled_dot_product_attention(query, key, value, is_causal=is_causal),
+        torch.nn.functional.scaled_dot_product_attention(query, key, value, attn_mask=attn_mask, is_causal=is_causal),
+        trident.function.scaled_dot_product_attention(query, key, value, attn_mask=attn_mask, is_causal=is_causal),
     )
 
 
 @pytest.mark.parametrize(
-    "num_batches, num_heads, y_size, x_size, is_causal", [(4, 8, 128, 64, True), (4, 8, 128, 64, False)]
+    "num_batches, num_heads, y_size, x_size, is_causal, with_bias",
+    [(4, 8, 128, 64, False, False), (4, 8, 128, 64, True, False), (4, 8, 128, 64, False, True)],
```
Comment (daemyung) [inline tests/test_attention.py]: `with_bias`를 새로 만들지 말고 `attention_mask`를 위한 테스트를 `test_backward`에 추가로 작성하는게 더 좋아보입니다. 코드가 조금 늘어나지만 분기문이 없어져서 테스트의 의도를 더 명확하게 파악할 수 있다고 생각합니다.

File: tests/test_attention.py [hunk 2]
```diff
@@ -20,55 +20,73 @@
...
 )
-def test_forward(num_batches, num_heads, y_size, x_size, is_causal, device):
+def test_forward(num_batches, num_heads, y_size, x_size, is_causal, with_bias, device):
     query = torch.randn(num_batches, num_heads, y_size, x_size, device=device)
     key = torch.randn_like(query)
     value = torch.randn_like(query)
+    attn_mask = torch.randn(num_batches, num_heads, y_size, y_size, device=device) if with_bias else None
```
Comment (daemyung) [inline tests/test_attention.py]: 변수명을 줄여쓰지 않았으면 좋겠습니다. `attention_mask`나 `mask`가 더 좋을 것 같습니다. `mask`를 쓰더라도 어텐션 테스트의 마스크이기 때문에 누구나 어텐션 마스크라고 생각할 것입니다.

File: trident/function/function.py [hunk 3]
```diff
@@ -230,6 +230,7 @@ def scaled_dot_product_attention(
     query: torch.Tensor,
     key: torch.Tensor,
     value: torch.Tensor,
+    attn_mask: torch.Tensor = None,
```
Comment (daemyung) [inline trident/function/function.py:233]: 여기는 파이토치와 동일해야하기 때문에 변경하지 않으셔도 됩니다.

File: trident/kernel/attention.py [hunk 4]
```diff
@@ -36,11 +36,15 @@ def forward(
         query_ptr: tl.tensor,
         key_ptr: tl.tensor,
         value_ptr: tl.tensor,
+        attn_mask_ptr: tl.tensor,
```
Comment (daemyung) [inline trident/kernel/attention.py]: `attention_mask`나 `mask`로 변경하는게 더 좋을 것 같습니다. 그리고 `attn_mask_ptr`은 `attn_mask_*_stride`와 같이 있는게 좋지 않을까요? `query_ptr, key_ptr, value_ptr`은 `y_size, x_size, ...`와 같은 정보를 가지고 있어서 이렇게 배치했습니다.

File: trident/kernel/attention.py [hunk 5]
```diff
@@ -131,6 +149,8 @@ def forward(
             output += language.dot(beta.to(dtype), value, use_accelerator, dtype)
             key_block_ptr = tl.advance(key_block_ptr, (0, y_block_size))
             value_block_ptr = tl.advance(value_block_ptr, (y_block_size, 0))
+            if attn_mask_ptr is not None:
```
Comment (daemyung) [inline trident/kernel/attention.py]: `if`문 앞에 줄바꿈이 있으면 보기 편할 것 같습니다.

**Review by daemyung — CHANGES_REQUESTED** [review]

File: tests/test_attention.py (hunk 2 above)
Comment (mejai1206) [reply tests/test_attention.py]: `mask`의 `shape`이 `(num_batches, num_heads, y_size, y_size)` 인데 마지막은 `x_size`가 아닌 `y_size`인 이유가 있을까요?

Comment (kakao-steve-ai) [reply tests/test_attention.py]: 공식을 간략히 쓰면, `Softmax(Q*K^T + mask)` 인데요. 현재는 `Q`와 `K`의 `shape`이 `(_, _, y_size, x_size)` 로 동일한 케이스만 고려하고 있습니다. `Q*K^T` 를 하면 `(y_size, x_size) * (x_size, y_size) = (y_size, y_size)` 가 나오게 됩니다. 그래서 `mask` 는 `shape` 이 `(_, _, y_size, y_size) `이어야 합니다.

File: tests/test_attention.py (hunk 1 above)
Comment (kakao-steve-ai) [reply tests/test_attention.py]: 테스트 수정하였습니다.

File: trident/kernel/attention.py (hunk 4 above)
Comment (kakao-steve-ai) [reply trident/kernel/attention.py]: `mask_ptr`과 `mask_*_stride`를 같이 배치하도록 수정하였습니다.

File: tests/test_attention.py [hunk 6]
```diff
@@ -81,3 +107,11 @@ def test_attention(num_batches, num_heads, y_size, x_size, is_causal, device, dt
     assert key.grad.dtype == dtype
     assert value.grad is not None
     assert value.grad.dtype == dtype
+
+    mask = torch.randn(num_batches, num_heads, y_size, y_size, **factory_kwargs, requires_grad=True)
+
```
Comment (daemyung) [inline tests/test_attention.py]: 의도된 줄바꿈일까요? 텐서 생성이라 밑에 코드랑 같이 있어도 될 것 같아요.

**Review by daemyung — CHANGES_REQUESTED** [review]
지금의 커밋 메세지는 새로운 Attention with bias를 구현한 것으로 생각되요. `Add support for bias in Attention`로 변경하면 어떨까요? 기존의 Attention에 Bias를 지원하는 의미니까 더 적절하다고 생각되요.

File: tests/test_attention.py (hunk 6 above)
Comment (kakao-steve-ai) [reply tests/test_attention.py]: 줄바꿈 삭제하였습니다.

**Comment by kakao-steve-ai** [comment]
> 지금의 커밋 메세지는 새로운 Attention with bias를 구현한 것으로 생각되요. `Add support for bias in Attention`로 변경하면 어떨까요? 기존의 Attention에 Bias를 지원하는 의미니까 더 적절하다고 생각되요.

Add, support가 중복되는거 같아 `Support bias in Attention` 로 수정하였습니다.

**Review by daemyung — APPROVED** [review]
LGTM

**Review by mejai1206 — APPROVED** [review]
LGTM
